### PR TITLE
Implements custom hotkeys

### PR DESCRIPTION
**Pull request by psykzz**
## About The Pull Request

Rebuilds on #1553 &#1558 to implement hotkeys as datums instead. 

## Changelog
:cl:
add: Custom hotkeys are now available, set them in the setup character screen.
tweak: Typical hotkeys are set to their defaults, but this can now be replaced.
add: Hotkey for dropping weeds as a Xeno added. Default Key: V
/:cl:

### CODE REVIEW

File: code/__HELPERS/sanitize_values.dm [hunk 1]
```diff
@@ -16,6 +16,9 @@
 	if(default)			return default
 	if(List && List.len)return List[1]
 
+/proc/sanitize_islist(value, default)
+	if(value && length(value) && islist(value))	return value
```
Comment (DominikPanic) [inline code/__HELPERS/sanitize_values.dm]: Just length(value) is enough (no need to double check, it if it's null length wont' pass)
Also, single line control statement

File: code/__HELPERS/sanitize_values.dm [hunk 2]
```diff
@@ -16,6 +16,9 @@
...
+	if(default)									return default
```
Comment (DominikPanic) [inline code/__HELPERS/sanitize_values.dm]: Single line control statement

File: code/modules/keybindings/bindings_client.dm [hunk 3]
```diff
@@ -45,11 +28,15 @@
 	if(!(next_move_dir_add & movement))
 		next_move_dir_sub |= movement
 
+	for (var/datum/keybinding/kb in prefs.key_bindings[_key])
```
Comment (DominikPanic) [inline code/modules/keybindings/bindings_client.dm]: You could skip the typecheck if you made sure only that type is present there.

File: code/modules/keybindings/bindings_client.dm [hunk 4]
```diff
@@ -11,26 +11,9 @@
...
-		if("F4")
-			get_msay()
-			return
-		if("F5")
-			get_dsay()
-			return
-		if("F12") // Toggles minimal HUD
-			mob.button_pressed_F12()
-			return
+	for (var/datum/keybinding/kb in prefs.key_bindings[_key])
```
Comment (DominikPanic) [inline code/modules/keybindings/bindings_client.dm]: Same here

File: code/__HELPERS/sanitize_values.dm (hunk 1 above)
Comment (psykzz) [reply code/__HELPERS/sanitize_values.dm]: Following the convention of the file here (with the single line control statement)

File: code/datums/keybinding/carbon.dm [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    
+/datum/keybinding/carbon/toggle_throw_mode
+    key = "R"
+    name = "toggle_throw_mode"
+    full_name = "Toggle throw mode"
+    description = "Toggle throwing the current item or not."
+    category = CATEGORY_CARBON
+    
+/datum/keybinding/carbon/toggle_throw_mode/down(client/user)
+    to_chat(user, "throwing test")
```
Comment (psykzz) [inline code/datums/keybinding/carbon.dm]: my bad debug code

**Comment by psykzz** [comment]
@Rohesie save to trade merge now, ive fixed the issues with preferences being unbound.

File: code/modules/client/preferences_savefile.dm [hunk 6]
```diff
@@ -19,6 +19,9 @@
 					fdel(delpath)
 				break
 		return FALSE
+    
+  if(savefile_version < 26)
+		S["key_bindings"] << deepCopyList(GLOB.keybinding_list_by_key)
```
Comment (Rohesie) [inline code/modules/client/preferences_savefile.dm]: Is this properly indented? Travis is complaining:
`code/modules/client/preferences_savefile.dm:23:warning: if: if statement has no effect`

Comment (psykzz) [reply code/modules/client/preferences_savefile.dm]: @Rohesie fixed, it was some weird indentation from the merge conflicts

**Comment by Rohesie** [comment]
H Y P E

**Comment by Rohesie** [comment]
Tested, it works.

**Pull request merged**